stdlib/asm/crypto/fri.masm

```diff
@@ -0,0 +1,873 @@
+#! I say partially because it can only be applied when index j is odd.
+#!
+#! Expected stack top
+#!
+#! [a3, a2, ⍳1, ⍳0, _, _, ν1, ν0, ...]
+#!
+#! Final stack state
+#!
+#! [_, _, ν1', ν0', ...] | ν' = (ν * a) + ⍳ s.t. a = (a3, a2), ⍳ = (⍳1, ⍳0), ν = (ν1, ν0), ν' = (ν0', ν1')
+proc.accumulate_into_numerator_odd_index
+    ext2mul
+    swap
+    movup.5
+    add
+    swap
+    movup.4
+    add
+    movdn.3
+    movdn.3
+end
+
+#! Given remainder codeword with 64 evaluations ( as starting memory address on stack ), this routine attempts to
+#! probabilistically verify remainder following the technique described in https://github.com/0xPolygonMiden/miden-vm/issues/592
+#!
+#! Expected stack state
+#!
+#! [mem_ptr, ...]
+#!
+#! A few assumptions about mem_ptr
+#!
+#! - mem_ptr should be absolute memory address of the beginning of remainder codeword
+#! - Each evaluation is 2 field elements wide because they belong to quadratic extension field over F_p | p = 2^64 - 2^32 + 1
+#! - Meaning each memory address will hold two consecutive evaluations
+#! - Words ( four field elements ), in memory, are laid out in this order (a1_0, a1_0, a0_1, a0_0)
+#! - That means (a0_1, a0_0) -> first evaluation and (a1_1, a1_0) -> next evaluation
+#! - Next 31 memory addresses should be holding remaining 62 evaluations
+#! - If mem_ptr holds (a1_0, a1_0, a0_1, a0_0), then mem_ptr + 1, must hold (a3_0, a3_0, a2_1, a2_0)
+#! - That way, mem_ptr + 31 should be holding (a63_0, a63_0, a62_1, a62_0)
+#!
+#! Final stack looks like
+#!
+#! [...]
+#!
+#! If remainder verification fails, execution of the program stops.
+export.verify_remainder_64.4
+    push.64
+    push.8
+
+    adv.ext2intt
+
+    drop
+    drop
+    dup
+
+    # Initialize Rescue state
+    push.144.0.0.0.0.0.0.0.0.0.0.0
+
+    # Consume `p` into Rescue state
+    repeat.16
+        mem_stream
+    end
+
+    movup.12
+    drop
+    locaddr.0
+    movdn.12
+
+    # Consume `q` into Rescue state
+    repeat.2
+        adv_pipe
+    end
+
+    movup.12
+    drop
+
+    dropw
+    swapw
+    dropw
+    drop
+    drop
+
+    # τ on stack top
+    # Compute β
+
+    push.0.0.0.0 # accumulators ( denominator, numerator in order )
+
+    # for j = 0
+    push.1
+    exec.compute_common_factor_of_beta
+    exec.accumulate_into_denominator_even_index
+
+    dup.8
+    push.0.0.0.0
+    movup.4
+    mem_loadw
+
+    exec.accumulate_into_numerator_even_index
+
+    # for j = 1
+    push.549755813888
+    exec.compute_common_factor_of_beta
+    exec.accumulate_into_denominator_odd_index
+    exec.accumulate_into_numerator_odd_index
+
+    # for j = 2
+	push.70368744161280
+	exec.compute_common_factor_of_beta
+    exec.accumulate_into_denominator_even_index
+
+    dup.8
+    add.1
+    push.0.0.0.0
+    movup.4
+    mem_loadw
+
+    exec.accumulate_into_numerator_even_index
```

Comment:
I would also try to combine these into a single procedure call. Something like `accumulate_beta_even`. I think one extra issue here is that we have custom code for loading the next set of coefficients for each value of $j$. But I think we can make the code general as follows:
```
movup.8
add.1
movdn.8
push.0.0.0
dup.11
push.0
swap
loadw
```
The above should actually take 1 less cycle as compared to the code we have now because `dup.8` takes 3 cycles while `dup.11` takes just 1 cycle.

To make the above work, the initial address in slot 8 should actually start out at 1 less than the actual address (because we first increment and then load). So, it will require one subtraction somewhere upfront.

Reply:
Oh this is interesting. I can try to implement this.